monai/transforms/inverse.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import warnings
+from typing import Dict, Hashable, Optional, Tuple
+
+import numpy as np
+import torch
+
+from monai.transforms.transform import RandomizableTransform, Transform
+from monai.utils import optional_import
+
+sitk, has_sitk = optional_import("SimpleITK")
+vtk, has_vtk = optional_import("vtk")
+vtk_numpy_support, _ = optional_import("vtk.util.numpy_support")
+
+__all__ = ["InvertibleTransform", "NonRigidTransform"]
+
+
+class InvertibleTransform(Transform):
+    """Classes for invertible transforms.
+
+    This class exists so that an ``invert`` method can be implemented. This allows, for
+    example, images to be cropped, rotated, padded, etc., during training and inference,
+    and after be returned to their original size before saving to file for comparison in
+    an external viewer.
+
+    When the `__call__` method is called, a serialization of the class is stored. When
+    the `inverse` method is called, the serialization is then removed. We use last in,
+    first out for the inverted transforms.
+
+    Note to developers: When converting a transform to an invertible transform, you need to:
+        1. Inherit from this class.
+        2. In `__call__`, add a call to `append_applied_transforms`.
+        3. Any extra information that might be needed for the inverse can be included with the
+            dictionary `extra_info`. This dictionary should have the same keys regardless of
+            whether `do_transform` was True or False and can only contain objects that are
+            accepted in pytorch's batch (e.g., `None` is not allowed).
+        4. Implement an `inverse` method. Make sure that after performing the inverse,
+            `remove_most_recent_transform` is called.
+    """
+
+    def append_applied_transforms(
+        self,
+        data: dict,
+        key: Hashable,
+        extra_info: Optional[dict] = None,
+        orig_size: Optional[Tuple] = None,
+    ) -> None:
+        """Append to list of applied transforms for that key."""
+        key_transform = str(key) + "_transforms"
+        info = {
```

Comment:
Should we use constant values for these keys as they are used in every transform?
Maybe self.ID, self.ORIG_SIZE, self.EXTRA_INFO?